cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -400,15 +413,166 @@ async function activateOption(option, loadBalancerDNS, loadBalancerHostedZone) {
+    // Look for validation options that will no longer needed by this service.
+    const certificates = await serviceCertificates();
+    const { certPendingDeletion, certInUse } = categorizeCertificates(aliases, certificates);
+    let optionsPendingDeletion = await unusedOptionsByService(certPendingDeletion, certInUse);
+
+    // For each of the options pending deletion, validate if it is in use by other services. If it is, Copilot
+    // will not delete it.
+    let promises = [];
+    for (const option of optionsPendingDeletion) {
+        const domainName = option["DomainName"];
+        const {route53Client} = await domainResources(domainName);
+        const promise = inUseByOtherServices(loadBalancerDNS, domainName, route53Client).then((isUsed) => {
+            if (isUsed) {
+                optionsPendingDeletion.delete(option);
+            }
+        });
+        promises.push(promise);
+    }
+    await Promise.all(promises);
+    return optionsPendingDeletion;
+}
+
+/**
+ * Retrieve all certificates used for the service and cache the results.
+ * @returns {Array<Object>} An array of descriptions for the certificates used by the service.
+ */
+async function serviceCertificates() {
+    if (cachedServiceCertificates) {
+        return cachedServiceCertificates;
+    }
+    let { ResourceTagMappingList } = await clients.resourceGroupsTaggingAPI().getResources({
+        TagFilters: [
+            {
+                Key: "copilot-application",
+                Values: [appName],
+            },
+            {
+                Key: "copilot-environment",
+                Values: [envName],
+            },
+            {
+                Key: "copilot-service",
+                Values: [serviceName],
+            }
+        ],
+        ResourceTypeFilters: ["acm:certificate"]
+    }).promise();
+
+    cachedServiceCertificates = [];
+    let promises = [];
+    for (const {ResourceARN: arn} of ResourceTagMappingList) {
+        let promise = clients.acm().describeCertificate({
+            CertificateArn: arn
+        }).promise().then(( { Certificate } ) => {
+            cachedServiceCertificates.push(Certificate);
+        });
+        promises.push(promise);
+    }
+    await Promise.all(promises);
+    return cachedServiceCertificates;
+}
+
+/**
+ * Retrieve the validation options that are pending deletion. An option is pending deletion if it is only used to
+ * validate a certificate that is pending deletion.
+ * @param {Array<Object>} certsPendingDeletion
+ * @param {Array<Object>} certsInUse
+ * @returns {Promise<Set<Object>>} options that are pending deletion.
+ */
+async function unusedOptionsByService(certsPendingDeletion, certsInUse) {
+    let optionsPendingDeletion = new Map();
+    for (const { DomainValidationOptions: validationOptions } of certsPendingDeletion) {
+        for (const option of validationOptions) {
+            optionsPendingDeletion.set(JSON.stringify(option), option);
```

Comment:
Do you know if there are any fields that could be problematic here within validation options, like `ValidationStatus`?

I think what I'm trying to ask is instead of the entire JSON object is there a better way of telling that two validation options have the same intention? maybe just looking at `ResourceRecord`?

Reply:
Good call out! Updated the code by just looking at `ResourceRecord`. 

We only manage `ResourceRecord` (not email); therefore, the `option` in this script is pretty much a synonymous to "the resource record of a validation option".

Therefore, it's correct to use `ResourceRecord` as the identifier, as well as only taking the option into consideration if it has a `ResourceRecord` field.